packages/buckets/src/api/index.ts

```diff
@@ -542,42 +578,63 @@ export async function bucketsPushPath(
             },
             root: event.root?.path ?? '',
           }
-          resolve(res)
+          return resolve(res)
         } else if (opts?.progress) {
           opts.progress(event?.bytes)
         }
       } else {
-        reject(new Error('Invalid reply'))
+        return reject(new Error('Invalid reply'))
       }
     })
     client.onEnd((code, msg) => {
       if (code !== grpc.Code.OK) {
         const message = msg ? msg : code.toString()
-        reject(new Error(message))
+        return reject(new Error(message))
       } else {
-        resolve()
+        return resolve()
       }
     })
     if (source) {
       const head = new PushPathRequest.Header()
       head.setPath(source.path || path)
       head.setKey(key)
+      // Setting root here ensures pushes will error if root is out of date
+      let root = ''
+      if (opts?.root) {
```

Comment:
The new behavior is that we "trust" the caller if they explicitly provide a root, but otherwise, make another API call to check the current known root.